Make background cube models

mapazarr opened the pull request:
Adding a PRELIMINARY 1st working version of the script to produce bg cube models. This is a continuation of the work started in: https://github.com/gammasky/hess-host-analyses/pull/1.
TODO list:
- [x] divide the code into smaller functions
- [x] move most of the functionality to Gammapy classes/functions
- [x] implement H.E.S.S. to Gammapy obs lists converter
- [x] reorganize the bg models class
- [x] add simple dummy bg event list generator
- [x] add tests for scripts and new methods in Gammapy
- [x] add script to compare 2 sets of bg cube models
- [x] remove DEBUG plots
- [x] implement comments from: https://github.com/gammasky/hess-host-analyses/pull/1 (and this PR)
- [x] update/implement docs (stringdocs and high-level docs)

mapazarr commented:
I'm aware of the travis-ci failure. It's because of the use of `matplotlib` without marking it as optional dependency (using the proper decorator in the tests). I'll work on that. It'll get fixed once I remove the plotting part (used for DEBUG for now).

mapazarr commented:
@cdeil I will need some fake data (event lists) for the tests. Is there something available? Otherwise I would write a supersimple random tool to do so.

mapazarr commented:
I'm aware of the travis-ci failure. It's because of the use of `scipy` without marking it as optional dependency (using the proper decorator in the tests). I'll work on that.

cdeil commented:
@mapazarr – A simple tool that samples background events according to some simple shape and spectrum would be good, even if it takes a day (but not more) to build. This way you'll have an easy way to generate example data, and you'll have "true" background cube models that you can compare the "reconstructed" ones against.

mapazarr commented:
I'm aware that I need to rebase. Since the PR is still WIP, I'll do this later.

mapazarr commented:
@cdeil I have divided the bg cube model production into smaller functions and added them to Gammapy. (I still need to do it for the observation grouping.) Can you please have a look at the structure of the code in `background/make.py` and tell me if you like it?

mapazarr commented on gammapy/background/make.py:
```diff
@@ -0,0 +1,470 @@
+#           'set_zero_level',
+#           'smooth'
+# @ C. Deil: does it make sense?
+
```
@cdeil What do you think about this TODO?

cdeil replied:
Yes, I think having separate container classes makes sense:
- one that represents a cube
- one that contains the counts, time and rate cubes

I don't have a clear picture yet on whether it would be good to have other classes, e.g.
- Grouping specification
- The algorithm that takes a grouping specification and produces all the cubes

If you can come up with a written down proposal for the hangout on Wednesday that would be good,
otherwise we start by reviewing this PR and then do some brainstorming how to best structure this (within the given time left, ~ 1 week).

mapazarr replied:
OK, separated.

cdeil commented:
Please rebase on master now (making a backup copy before you start).
This has several advantages:
- travis-ci runs
- I can more easily try out your branch
- sometimes the diff on github is more accurate
- one large rebase at the end is harder than smaller rebases with only few conflicts from time to time.

cdeil commented on gammapy/background/make.py:
```diff
@@ -0,0 +1,470 @@
+        Observation list to use for the histogramming.
+    fits_path : str
+        Path to the data files.
+    events_cube : `~gammapy.background.CubeBackgroundModel`
```
Please rename `events_cube` -> `counts_cube` everywhere.

mapazarr replied:
Done.

cdeil commented on gammapy/background/make.py:
```diff
@@ -0,0 +1,470 @@
+        Cube container for the events.
+    livetime_cube : `~gammapy.background.CubeBackgroundModel`
+        Cube container for the livetime.
+    DEBUG : int
```
This is not a good way to do logging.

Please change to this way of generating log messages everywhere:
https://gammapy.readthedocs.org/en/latest/development/index.html?highlight=logging#generating-log-messages

mapazarr replied:
Fixed.

cdeil commented on gammapy/background/make.py:
```diff
@@ -0,0 +1,470 @@
+    return events_cube, livetime_cube
+
+
+def divide_bin_volume(cube):
```
Such a function that just operates on one object is better located as a method on the object.
Any reason not to make this a `CubeBackgroundModel` member function?

mapazarr replied:
Moved to `Cube`.

cdeil commented on gammapy/background/make.py:
```diff
@@ -0,0 +1,470 @@
+    return cube
+
+
+def set_zero_level(cube):
```
Same here ... why not make this a cube member function?
Is it because you are thinking about code re-use with other cube objects (like `SpectralCube`) that could also use this function?

mapazarr replied:
Moved to `Cube`.

cdeil commented on gammapy/background/make.py:
```diff
@@ -0,0 +1,470 @@
+
+    Smooth method:
+
+    1. slice model in energy bins: 1 image per energy bin
```
If the smoothing happens on 2D images, it's best (for readability and code re-use) to split that part that operates on 2D images out into a separate function. You could put it in `gammapy/image/utils.py`.

mapazarr replied:
Sorry, no time for this right now. Since we anyway are not sure this is the best smoothing method, I'm not sure it is worth the effort. Maybe in a later PR.

cdeil commented on gammapy/background/make.py:
```diff
@@ -0,0 +1,470 @@
+        Observation list to use for the histogramming.
+    fits_path : str
+        Path to the data files.
+    DEBUG : int
```
Again ... it's important to use Python logging.

This function could be one of those cases I'm talking about in the last paragraph [here](https://gammapy.readthedocs.org/en/latest/development/index.html#generating-log-messages) where it makes sense that they optionally take their own logger as input. [Example](https://gammapy.readthedocs.org/en/latest/api/gammapy.data.EventListDatasetChecker.html).

mapazarr replied:
Done.

cdeil commented on gammapy/scripts/make_bg_cube_models.py:
```diff
@@ -0,0 +1,329 @@
+def create_bg_observation_list(fits_path):
+    """Make total observation list and filter the observations.
+
+    In a first version, all obs taken within 3 deg of a known source will be rejected. If a source is extended, twice the extension is added to the corresponding exclusion region radius of 3 deg.
```
Break line so that it's easily readable.

mapazarr replied:
Done.

cdeil commented on gammapy/scripts/make_bg_cube_models.py:
```diff
@@ -0,0 +1,329 @@
+# https://github.com/mapazarr/gammapy/blob/bg-api/dev/background-api.py#L55
+altitude_edges = Angle([0, 20, 23, 27, 30, 33, 37, 40, 44, 49, 53, 58, 64, 72, 90], 'degree')
+if DEBUG > 1:
+    altitude_edges = Angle([0, 45, 90], 'degree')
```
I think we and others will want to try many different ways of grouping the off runs, and binning / smoothing / fitting the cubes. So having one hard-coded config and one alternative "debug" config for quicker tests doesn't scale to this task.

Making this nice and easy to configure is a hard problem and we have to think / discuss how to do this best.
My first thought would be do have some "...Config" or "...Specification" class (or plain Python dict for now)
and then for the command line tool accept a YAML config file describing this configuration as input.

I think that's what I want to do in Gammapy for all the configurable analyses ... YAML config files.
There's many packages that do this ... here's just one example: http://fermipy.readthedocs.org/en/latest/quickstart.html#creating-a-configuration-file

cdeil replied:
@mapazarr – For now you could go one step in this direction and collect the parameters for the "hess-mayer" and "hess-debug" config in a dict and let the user select which config to run via a string option.

mapazarr replied:
Superdone: `ObservationGroups` class.

cdeil commented on gammapy/scripts/make_bg_cube_models.py:
```diff
@@ -0,0 +1,329 @@
+
+            # save the observation list to a fits file
+            outfile = outdir +\
+                     'bg_observation_table_alt{0}_az{1}.fits.gz'.format(i_alt, i_az)
```
This way of naming files doesn't scale to other grouping parameters or schemes.
Maybe the best would be to just use GROUP_ID here and then have a separate data structure that defines the grouping and can be written / read.

(I'm not claiming to have a concrete great suggestion here ... this is the issue we've been discussing for a long time now that it's hard to have a good API / classes for observation grouping.)

mapazarr replied:
Done. It's hard though to compare models. See workaround in  `examples/plot_bg_cube_model_comparison.py` (included in this PR). Lines ~50 - ~70.

cdeil commented on gammapy/scripts/make_bg_cube_models.py:
```diff
@@ -0,0 +1,329 @@
+        for oldfile in os.listdir(outdir):
+            os.remove(outdir + oldfile)
+
+    # loop over altitude and azimuth angle bins: remember 1 bin less than bin boundaries
```
Again, this function only works for exactly this grouping in `(azimuth, zenith)`.
It would be much more flexible and better to have one for loop over observation groups here, i.e. to have the observation grouping and background modeling decoupled!

mapazarr replied:
Again: done.

cdeil commented on setup.py:
```diff
@@ -94,6 +93,7 @@
     'gammapy-iterative-source-detect = gammapy.scripts.iterative_source_detect:main',
     'gammapy-look-up-image = gammapy.scripts.look_up_image:main',
     'gammapy-model-image = gammapy.scripts.model_image:main',
+    'gammapy-make_bg_cube_models = gammapy.scripts.make_bg_cube_models:main',
```
Please stick with the naming scheme to always use hyphens, not underscores in command line tools.
Hyphens are easier to type (no shift) and mixing hyphens and underscores will mean that users always have to guess.
`gammapy-make_bg_cube_models` -> `gammapy-make-bg-cube-models`

mapazarr replied:
Fixed.

cdeil commented on gammapy/background/__init__.py:
```diff
@@ -10,3 +10,4 @@
 from .template import *
 from .kernel import *
 from .models import *
+from .make import *
```
I think make is too generic a name, plus it would be good to have the class and functions operating on that class (if any remain standalone) in the same file.
So how about renaming `make.py` -> `cube_model.py` and moving the `CubeBackgroundModel` class here?

mapazarr replied:
Well, most of the functions should probably move to other places. I like having the generic naming of `models.py` and `make.py`, in case other models come in the future and can share the same files. I don't have a strong opinion on that. I was just following the same convention as in the `dataset` module.

After this clarification: should I keep it as it is? or should I use `cube` specific naming?

BTW, the generic `Cube` class mentioned in https://github.com/gammapy/gammapy/pull/319#discussion-diff-36716979 could go into a `cube.py` file, similar to `maps.py`.

cdeil replied:
I especially don't like `make.py`, because a year from now there will be lots of other code that makes things in `gammpay.background`, e.g. reflected regions or exclusion regions or ... and that code shouldn't be in the same file as what's in `make.py` now.

So either `cube_model.py` like I suggested, or `cube.py` like you did seems better to me.

But in the end ... we will accumulate functionality for a while and then try to do a consistent naming across Gammapy before the 1.0 release. So it's not terribly important what we choose now ... you decide!

mapazarr replied:
Done. I moved all methods to either `Cube` in `cube.py` or `CubeBackgroundModel` in `models.py`. I left only the skeleton in `make_bg_cube_model` in `make.py`.

cdeil commented:
@mapazarr – I've left a bunch of inline comments. Some (like rebase, logging) are straightforward to address, others (like grouping, configuring) are vague and need more thought / discussion.

One thing that would help me very much to give useful feedback on this would be if I had a high-level docs page with examples how to run the command line tool so that I can quickly start playing around with this. I'm not sure if generating a ~ 10 run example dataset and using it for a public docs page in Gammapy is best, or a HESS-internal page showing how to do it for a small and quick and the full HESS off run dataset. Ideally both. :-)

mapazarr commented:
@cdeil OK, I added a documentation page. Sorry, I thought it was clear (for now), when running the inline command with `--help` and left the doc for later. It's a subpage of the main `background` page.

I still didn't have time to work on the test data for Gammapy. As for now, it runs on H.E.S.S. data (prod02). You can run a quick test with:
```
gammapy-make_bg_cube_models /path/to/fits/event_lists/base/dir --test True
```
The whole production (without `--test`) takes ~ 5 min on my computer.

cdeil commented on docs/background/make_models.rst:
```diff
@@ -0,0 +1,53 @@
+
+The output files are created in the current directory:
+
+* ``bg_observation_table.fits.gz``: total observatin table used for
```
Typo: observatin -> observation

mapazarr replied:
Fixed.

cdeil commented on docs/background/make_models.rst:
```diff
@@ -0,0 +1,53 @@
+
+The ``gammapy-make_bg_cube_models`` tool has a few options. For a full list of options, please use:
+
+  .. code-block:: bash
```
Remove space before the `..` This doesn't need to be indented:
![screen shot 2015-08-12 at 11 40 10](https://cloud.githubusercontent.com/assets/852409/9220988/eadcf5de-40e6-11e5-8817-4e84ebfc23aa.png)

mapazarr replied:
Done

cdeil commented:
I'm seeing this error from ftverify for the cube model image FITS header:
https://gist.github.com/cdeil/8665a22897c6ebe71680#file-gistfile1-txt-L1
If it's not clear how to fix it, please put it on the agenda for our hangout tonight.

cdeil commented:
I ran the test example and it worked. :clap: 

[log](https://gist.github.com/cdeil/bf0ca379a959f68eb416) and generated files:
```
$ find .
.
./bg_cube_models
./bg_cube_models/bg_cube_model_alt0_az0_image.fits.gz
./bg_cube_models/bg_cube_model_alt0_az0_table.fits.gz
./bg_cube_models/bg_cube_model_alt1_az0_image.fits.gz
./bg_cube_models/bg_cube_model_alt1_az0_table.fits.gz
./bg_observation_table.fits.gz
./splitted_obs_list
./splitted_obs_list/bg_observation_table_alt0_az0.fits.gz
./splitted_obs_list/bg_observation_table_alt1_az0.fits.gz
```

A few suggestions for improvement based on this first impression:

* Don't be so chatty in the log (e.g. all the obs tables shouldn't be printed, or if you really want to, make this debug output.
* As already mentioned, use `group_XXXX` with the GROUP_ID instead of `alt0_az0` (scales to other grouping schemes).
* Probably best to put the group obs tables in the same folder (or even file as an extra extension) as the cubes.
* We need a way to store the obs grouping info on disk, so that it can be later used to associate observations with background models, like @mimayer when he creates a soft link for each obs to a background model. This is probably for a different script and PR, but this script has to store the required obs grouping info somehow. This is difficult and we can try to find a solution in the hangout tonight.

mapazarr commented:
@deil I'm aware of the problematic related to the observation grouping. I'm working on an `observation groups` class that stores the groups into an `~astropy.table.Table`. This can easily be saved to disk and read back.

![obs_groups_test](https://cloud.githubusercontent.com/assets/11326731/9233775/d0b31be0-40e9-11e5-874d-808da6a6cc6a.png)

cdeil commented:
@mapazarr – Great that you've started working on generalising the observation grouping!
A table will only work for min / max cuts, right?
I'm not sure this is general enough for the long run, but it's probably a very good choice to get something pretty general in the given short amount of time.

mapazarr commented:
@cdeil @adonath I added 2 new classes to deal with observation groups (still WIP).

mapazarr commented:
@cdeil @adonath Ok, I have a working version of the observation grouping classes/methods. Please have a look at the docs of `gammapy.obs.ObservationGroups` and `gammapy.obs.ObservationGroups.group_observation_table`. I still have to define the inline command tool, but this will come later.

mapazarr commented:
@cdeil I'm not sure what you mean by "only min max cuts" in your previous comment. In principle you can define any amount of axes with the desired binning in continuous or discrete `Quantity`-like variables and the algorithm combines all possibilities to define the groups.
What else did you have in mind?

mapazarr commented:
Ok, the inline command uses now the `ObservationGroups` class.

mapazarr commented:
@cdeil @adonath I added 2 functions to deal with observation table format conversion and integrated them into the `datastore` classes, so that the format of obs tables and datastore obs tables imported from the H.E.S.S. `runinfo.fits` file follow the convention here:
http://gammapy.readthedocs.org/en/latest/dataformats/observation_lists.html
(I also fixed small bugs in the datastore classes).

I first attempted to add the functions to the `datasets` module in an extra file `convert.py`, but them I had problems, when importing then within `obs.datastore.py`. (A circular dependency I think?) Before you ask, I had updated the corresponding `__init.py__` file. Please see this commit: https://github.com/mapazarr/gammapy/commit/675df3ec5317539bca312fa1d34bd09e9ede12ba

Moving the file to the `obs` module didn't help. So I moved them to `obs.datastore.py` and it worked. (It also works if they are in `obs.observation.py`, but I didn't like them there,  since the functions are specific to a particular experiment.)

Please let me know if you like the current location of the functions, otherwise, let me know what is the best location for them.

mapazarr commented:
@cdeil @adonath One possibility for the converter classes could be to add them as methods of `ObservationTable`, but I'm not sure they belong there. It seems they are more like utility functions, rather than class members.

mapazarr commented:
Another possibility would be to move `obs.datastore.py` to `datasets`. Since it deals with the dataset files, maybe it should go there?

mapazarr commented:
Ok, I am leaving the converters as they are for now. I filed issue #328 so any further discussion on this topic should go there.

mapazarr commented:
@cdeil @adonath I'm still working on this PR, and have still many things to do. However, the structure of the classes (in `obs`, `datasets` and `background`) is more or less final, except for small additions/modifications that arise from the open TODOs, and for issue #328, where I anyway would like your opinion.

So if you have time, you can have a look at the structure of the `bg cube`, `ovservation grouping` and `observation list converter` classes and comment. Please consider also that I won't have much time to implement many comments in the remaning time of GSoC.

Currently I am working on generalising the `Cube` container class, which was already quite general, except for a couple of hard coded variables, so it can potentialy be used as container for other classes in the future, like i.e. `SpectralCube`, or a future `ProjectedBackgroundCubeModel`.

cdeil commented:
> @cdeil I'm not sure what you mean by "only min max cuts" in your previous comment. In principle you can define any amount of axes with the desired binning in continuous or discrete Quantity-like variables and the algorithm combines all possibilities to define the groups.
What else did you have in mind?

What I had in mind was something fancy that we shouldn't attempt here, it's a large separate project. Implement a score function that computes for a pair of observations how "similar" they are in terms of background model. Input variables to this score function could be `n_tel`, `zenith_angle`, `n_events`, parameters computed from the background distribution, like energy quantiles or offset distribution quantiles. Such a score function could be used to find for each "on" observation the 10 "most similar" off observations and stack those to get an on-off background method without the issue of low stats and bad alpha / significance loss in the off run.
It could also be used to cluster the off runs into groups. In those cases one would probably like to serialise the score function somehow and this probably doesn't fit nicely in a table. That's why I had JSON or YAML in mind for storing the observation grouping spec. Those are hierarchical formats that basically support storing any info / parameters we like, in addition to allowing table info.

But using tables for now is fine, no need to change to a more complex format now for something more general that might or might not happen in the future.

cdeil commented:
@mapazarr – My suggestion here would be that you start wrapping this up, i.e. address all the little TODO's you mention in the code comments.

As you say there's not much time left, so even if I'd structure the code differently it's unlikely that there is time for you to do this. And after all: if it works, it works, and we can re-factor later.

What I'd like to avoid is merging code in master with a ton of little TODOs left, so please try to do those and let's start polishing this now.

cdeil commented on gammapy/scripts/tests/test_make_bg_cube_models.py:
```diff
@@ -0,0 +1,24 @@
+
+@pytest.mark.skipif('not HAS_SCIPY')
+@pytest.mark.xfail
+def test_make_bg_cube_models_main(tmpdir):
```
Implementing at least one test for this is really important -> please make this high priority.

mapazarr replied:
Done. I added a dummy dataset generator, otherwise the test would not work. Now the test is working, however, I'm not sure what the best strategy for the asserts is. Suggestions? Maybe we can discuss in the hangout.

mapazarr replied:
Done.

cdeil commented on gammapy/scripts/make_bg_cube_models.py:
```diff
@@ -0,0 +1,308 @@
+    fits_path : str
+        path to dir containing list of input fits event files
+    """
+    if DEBUG:
```
Just to be clear: I said before that you should use logging statements everywhere you have print statements.
But actually most of your print statements, like this, should just be removed IMO.

Please use your best judgement (maybe err on the side of leaving too many debug / info / warning statement) and then I'll leave inline comments which ones I think are useful to keep.

mapazarr replied:
I'm aware of this. I just didn't have time for now to deal with it. I'll do it soon.

mapazarr replied:
Done.

cdeil commented:
> Currently I am working on generalising the Cube container class, which was already quite general, except for a couple of hard coded variables, so it can potentialy be used as container for other classes in the future, like i.e. SpectralCube, or a future ProjectedBackgroundCubeModel.

I do *not* think that's a good idea to do this now. Re-structuring the Cube class so that it can share code with the SpectralCube class or other Gammapy classes is a noble idea, but I don't think it's feasible in the given limited time. It will need thought and discussion and then a dedicated pull request that refactors things.

You should try to wrap this up until the hangout on Wednesday. Then there will be some small improvements and polishing and adding extra tests or docs for the remaining two days, and then GSoC will be over.

mapazarr commented:
@cdeil there is a new test (appveyor), which actually failed. But it seems that there is a problem in the configuration of the test, rather than a problem in my commit, right? Should I worry about it?

cdeil commented:
Ignore testing on Windows on Appveyor. It's not properly set up yet.

mapazarr commented:
@cdeil @adonath the test for the inline commando tool for producing bg cube models is working now, though I'm not sure about the best strategy for the tests.

I also updated the documentation. Please have a look at the high-level docs generated from `docs/background/make_models.rst` and let me know your impressions.

Anyway, I'm pretty much done with the large additions of code. And what remains is clean-up, re-structure (split) the inline tool, add a few more options and eventually a few more tests. Lets discuss the strategy in the hangout today. I will start with code clean-up.

mapazarr commented:
@cdeil @adonath 
You can have a look at the docs here:
https://ms2.physik.hu-berlin.de/~mapaz/gammapy/docs/_build/html/
Especially, have a look at the high-level docs of the bg cube models production tool:
https://ms2.physik.hu-berlin.de/~mapaz/gammapy/docs/_build/html/background/make_models.html
and if you have time, test it.

cdeil commented:
@mapazarr – I don't have time to look at this before our hangout later today. We can start by you walking us through the code and docs. Please put the major questions you have for this in the hangout notes, and then we'll go through them and try to find solutions one by one.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -3,19 +3,26 @@
            'make_test_observation_table',
            'make_test_bg_cube_model',
+           'make_test_dataset',
+           'make_test_evenlist',
```
Typo: `evenlist` -> `eventlist` (missing "t")

mapazarr replied:
Fixed.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -356,6 +364,213 @@ def make_test_bg_cube_model(detx_range=Angle([-10., 10.], 'degree'),
+    else:
+        s_error = "Warning! Storage scheme for {}".format(observatory_name)
+        s_error += "not implemented. Only H.E.S.S. scheme is available."
+        raise RuntimeError(s_error)
```
This should be a `ValueError` (always use that to handle cases of invalid inputs).

mapazarr replied:
OK. I thought, im principle it should work for other schemes also, only we still don't have support for them, so it's not exactly a value error, but yes, it's also not completely a RuntimeError as well.
Changed.

cdeil commented on gammapy/datasets/tests/test_make.py:
```diff
@@ -72,9 +76,41 @@ def test_make_test_bg_cube_model():
+
+def test_make_test_dataset(tmpdir):
+    # create a dataset
+    fits_path = str(tmpdir.join('test_dataset'))
```
Why put everything in a sub-folder `test_dataset`.
You can just use `tmpdir` directly, no?

I.e. remove the line here and below write
```
make_test_dataset(fits_path=str(tmpdir),
```

mapazarr replied:
The idea is to put inputs and outputs in separate folders in order not to mix up things. So I'm not changing this.

cdeil commented on gammapy/obs/datastore.py:
```diff
@@ -278,3 +278,95 @@ def check_available_event_lists(self, logger=None):
+    described in :ref:`dataformats_observation_lists`.
+
+    Curretly only the H.E.S.S. scheme is supported, which follows the
+    guidelines dictated for CTA.
```
There are no guidelines for CTA yet ... remove that statement (or the whole sentence).

mapazarr replied:
OK, I thought there were from our conversations with Michi.
Removed.

cdeil commented on gammapy/obs/datastore.py:
```diff
@@ -278,3 +278,95 @@ def check_available_event_lists(self, logger=None):
+    ----------
+    obs_list : `~astropy.table.Table`
+        Observation list to convert.
+    scheme : str
```
```python
scheme : {'HESS'}
```

mapazarr replied:
Done.

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -128,6 +141,8 @@ def select_range(self, selection_variable, value_range, inverted=False):
         # build and apply mask
         mask = (value_range[0] <= value) & (value < value_range[1])
+        if np.allclose(value_range[0], value_range[1]):
+            mask = value_range[0] == value
```
I know it's equivalent, but I find adding extra parentheses makes the code a bit more readable:
```python
mask = (value_range[0] == value)
```

mapazarr replied:
Ok, but in a later comment you requested to avoid superfluous parentheses :-)
Done.

cdeil replied:
I know, but I think both decisions make sense.
Superfluous parentheses should only be inserted where it improves readability.
Here with `=` and `==` there's possibility to get confused what this statement does.
This is a borderline call, sometimes in complex expressions where operator precedence matters it's more important to insert "superfluous" parentheses, i.e. ones that are only relevant for humans.

Plus: I'm only allowed to make superficial, quick to address suggestions on this PR, so that's what I'm doing. :smile:

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -319,3 +339,569 @@ def select_observations(self, selection=None):
+
+    Examples
+    --------
+    Define an observation grouping:
```
I would find it really helpful to see the output of the print statements below.
My suggestion would be to:
- Move this example to the high-level docs page
- Link to it from this docstring
- Choose an example with fewer groups (~ 10) so that it's possible to print the relevant results.

mapazarr replied:
Done.

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -319,3 +339,569 @@ def select_observations(self, selection=None):
+
+    The class takes as input a list of `~gammapy.obs.ObservationGroupAxis`
+    objects and defines 1 group for each possible combination of the
+    bins defined in all axes. The groups are identified by a unique
```
The most common term for what you're describing here is "cartesian product":
https://en.wikipedia.org/wiki/Cartesian_product
https://docs.python.org/3.6/library/itertools.html#itertools.product

Can you use that term and maybe even link to the Wikipedia page?

mapazarr replied:
Ok, I added the term. The user can copy-paste it in google :-)

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -319,3 +339,569 @@ def select_observations(self, selection=None):
+                s +=','
+            s += ' ' + self.obs_group_axes[i_axis].name + ' = '
+            if self.obs_group_axes[i_axis].format == 'bin_edges':
+                s += '[' +\
```
Such multi-line expressions are very hard to read (and debug if there's an issue).
Please split it in separate statements:
```python
s += ...
s += ...
s += ...
```

mapazarr replied:
OK, I find them easier to read in some cases. But it's a matter of taste. Changed.

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -319,3 +339,569 @@ def select_observations(self, selection=None):
+            s += ' ' + str(self.obs_group_axes[i_axis].bins.unit)
+        return s
+
+    def print_axes(self):
```
Is this method really useful?
Probably should be removed?

mapazarr replied:
Done.

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -319,3 +339,569 @@ def select_observations(self, selection=None):
+        #print(self.info)
+        log.info(self.info)
+
+    def print_groups(self):
```
Is this method really useful?
Probably should be removed?

mapazarr replied:
Done.

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -319,3 +339,569 @@ def select_observations(self, selection=None):
+                elif format == 'bin_values':
+                   min_value = recover_units(self.obs_groups_table[name][i_row],
+                                              self.obs_groups_table[name])
+                   max_value = min_value
```
It's usually a good idea to end such if-elif-statements for multiple-choice options with an else clause that raises a ValueError if none of them matched. The reason is that if none of the options match, that can lead to weird errors further down or even worse, no error but incorrect output.

mapazarr replied:
I agree, but this is exactly the 1st thing in `ObservationGroupAxis.__init__`, so it shouldn't be necessary here. Otherwise I'd have to add it all over this and the Axis classes.

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -319,3 +339,569 @@ def select_observations(self, selection=None):
+        """Info string (str)"""
+        return ("{0} {1} {2}".format(self.name, self.format, self.bins))
+
+    def print(self):
```
Remove?
(it's rare to want to do specifically that, and in those cases the caller can just use the one line in this method)

mapazarr replied:
Done.

cdeil commented on gammapy/obs/tests/test_observation.py:
```diff
@@ -186,3 +194,90 @@ def test_select_sky_regions():
+                           ObservationGroupAxis('AZ', az, 'bin_edges'),
+                           ObservationGroupAxis('N_TELS', ntels, 'bin_values')]
+    obs_group = ObservationGroups(list_obs_group_axis)
+    assert ((0 <= obs_group.obs_groups_table['GROUP_ID']) &
```
Multi-line expressions are hard to read. Can't you just split this in two asserts and it'll be equivalent?

mapazarr replied:
It will be equivalent in the case of `AND`, but not in the case of `OR`. And I don't like splitting it because logically they belong together: it's an assert on the range of validity of one parameter. Splitting in 2 asserts it looks like it's 2 different tests, so it remains unchanged.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -356,6 +364,213 @@ def make_test_bg_cube_model(detx_range=Angle([-10., 10.], 'degree'),
+    """
+    # create output folder
+    outdir = fits_path
+    if not os.path.isdir(outdir):
```
This code shows up twice in this PR.
I haven't thought about if it's a good idea to do this, but if you think it is, OK to keep as is, but please split this out into a private utility function and call it from both places.

mapazarr replied:
Sorry, I can't think of a good place for this now. Moving to #333.

cdeil commented:
@mapazarr – I've left a few more inline comments ... all superficial and quick to address I think.

Can't wait to try this out!

mapazarr commented:
@cdeil @adonath I am using now the logger. Please have a look and make inline comments in the positions of the code, where you'd like to have changes.

mapazarr commented on docs/background/make_models.rst:
```diff
@@ -0,0 +1,121 @@
+                                    HESS bg_cube_models --test
+
+* Create background models using the method developped by Michael
+  Mayer (almost equal to the default case for now):
```
@cdeil should I add a link to Michi's thesis as a reference? I think it would be nice. Plus I don't feel like adding a long description about his method in the docs.

cdeil replied:
Yes, please link to https://publishup.uni-potsdam.de/frontdoor/index/index/docId/7150 . Definitely don't spent time describing the method ... I don't think it's something we want to keep (or at least not advertise via the public API) once one or two better grouping / smoothing methods are implemented.

mapazarr replied:
Done.

mapazarr commented on gammapy/scripts/tests/test_make_bg_cube_models.py:
```diff
@@ -0,0 +1,36 @@
+
+
+@pytest.mark.skipif('not HAS_SCIPY')
+@pytest.mark.parametrize("extra_options,something_to_test", [
```
@cdeil I'm not sure about the second element in the tuple for parametrize. In fact, I'm not sure at all we should parametrize this test:

* On the one hand: it would be nice to test several options. Especially in the future, when more options become available.
* On the other hand, each test takes a few seconds to run, and we don't want the tests to take forever, right?

Suggestions?

cdeil replied:
I'd say leave it parametrised. It's parametrised now, and as you say, we definitely want to test at least a few common options in the future.

Concerning the issue that the test is slow: ignore that for now.

In the future we can make it faster, e.g. by reducing number of events and bins in the cube.
And the input test dataset can be shared across test runs, i.e. only be generated once per test session.
But improving the tests like that takes time, i.e. is future work.

cdeil replied:
I forgot to mention the most important points for such high-level analysis tests: we should introduce a marker for "slow" or "analysis" tests that don't get run on travis-ci, but that we run before making a release or weekly or nightly on a Jenkins I have at MPIK.
These can run real analyses, and it's OK if they take many minutes and in total even hours ... as long as they finish within a ~ day that's no problem.

mapazarr replied:
OK, leaving unchanged.

mapazarr commented:
@cdeil I don't think is anything to worry about now, but I keep getting warnings because we define too long keywords in FITS files:
```
WARNING: VerifyWarning: Keyword name 'OBSERVATORY_NAME' is greater than 8 characters or contains characters not allowed by the FITS standard; a HIERARCH card will be created. [astropy.io.fits.card]
WARNING  VerifyWarning: Keyword name 'OBSERVATORY_NAME' is greater than 8 characters or contains characters not allowed by the FITS standard; a HIERARCH card will be created. [astropy]
WARNING: VerifyWarning: Keyword name 'TIME_FORMAT' is greater than 8 characters or contains characters not allowed by the FITS standard; a HIERARCH card will be created. [astropy.io.fits.card]
WARNING  VerifyWarning: Keyword name 'TIME_FORMAT' is greater than 8 characters or contains characters not allowed by the FITS standard; a HIERARCH card will be created. [astropy]
```
but maybe this is something to consider for the future.

mapazarr commented:
And I won another coffe! (the travis-ci icon, when the build succeeded is very similar to the starbucks one) :-)

I changed the logger messages a bit: I gave the user something to read while he waits that the models are done :-)

cdeil commented:
Concerning the > 8 character FITS header keywords pointed out by ftverify.
They should be replaced by <= 8 keywords, even if cryptic.

But as you way, this is not high priority, so if it's not quick to change here, OK to leave for later.

cdeil commented on gammapy/obs/datastore.py:
```diff
@@ -125,16 +124,16 @@ class DataStore(object):
         self.dir = dir
         self.index_table_filename = 'runinfo.fits'
         filename = os.path.join(dir, self.index_table_filename)
         print('Reading {}'.format(filename))
```
print -> log.info

mapazarr replied:
Done.

cdeil commented on gammapy/obs/tests/test_observation.py:
```diff
@@ -118,10 +123,12 @@ def test_select_time_box():
+                                                 random_state=random_state)
 
     # test box selection in time: (time_start, time_stop) within value_range
     print()
```
Remove print

mapazarr replied:
Done.

cdeil commented on gammapy/scripts/make_bg_cube_models.py:
```diff
@@ -0,0 +1,321 @@
+    test : bool
+        If true, run fast: skip many runs and catalog sources.
+    """
+    print()
```
No print statements allowed from Gammapy!
(makes it hard to impossible to configure what to do with the output for the user).

Like I said before: most of these print statements can be removed.
Some you think important you can keep, but turn them into log statements.

mapazarr replied:
Done. But for the record:
1) the info log statements look bad, especially with tables or empty lines.
2) You said print statements from command-line tools are ok.
I removed all for now.

cdeil replied:
Maybe log like this to avoid having the first line indented and add newlines to make it more readable if you want?
```python
# Just an example, not really meant to insert this here
log.info('\nObservation table:\n{}\n'.format(table))
```
This way we get nicer output, but the user still has control over where the log output goes.

If this is still not good for some reason, feel free to make a new issue so that we can try to find a better solution.

cdeil replied:
And concerning "print statements from command line tools are OK": I'd like all command line tools to be thin wrappers around library functions, that mostly do argument parsing. So long-term I'd like to make these functions more general and move them to appropriate places in Gammapy. Do they currently show up in the API docs of `gammapy.scripts`?

mapazarr replied:
Yes they do, shouldn't they?

Moving further logging issues to #333.

cdeil commented:
@mapazarr – Can you look through the inline comments at https://github.com/gammapy/gammapy/pull/319/files ... there's a few small ones that aren't addressed yet, but that I think are simple to address.

Also please open a new issue and move the parts of the task list at the top for which there is no time to address them in this PR to that issue.

There's no point in me doing another quick review and leaving more comments, because time's up, right?

cdeil commented on gammapy/background/cube.py:
```diff
@@ -0,0 +1,834 @@
+                u.Unit(data_unit, format='ogip')
+            # if it still fails, raise an exception
+            except ValueError:
+                raise ValueError("Cannot interpret units ({})".format(background_unit))
```
Undefined variable 'background_unit'

mapazarr replied:
Changed to 'Invalid unit format'.

cdeil commented on gammapy/background/make.py:
```diff
@@ -0,0 +1,106 @@
+        Observation list to use for the histogramming.
+    fits_path : str
+        Path to the data files.
+    a_la_michi : bool, optional
```
I'd prefer if you make this a multiple-choice option.
```python
method : {'default', 'michi'}
```
where default is the behaviour you currently have for `a_la_michi=False`.

(There's likely to be several methods implemented in the future and with multiple-choice it's extendable without changing the API.)

cdeil replied:
Probably the code becomes more readable if you have one large multiple-choice section:
```python
if method == 'default':
    .... do everything ...
elif method == 'michi':
    ... do everything ...
else:
    raise ValueError()
```
instead of three small if statements as you have now.
(But I'm not 100% sure, do what you think is best.)

mapazarr replied:
Done, done.

cdeil commented on gammapy/background/make.py:
```diff
@@ -0,0 +1,106 @@
+
+
+    ############################
+    # return the bg cube model #
```
Please remove the ascii-art comments. :art: :smile: 
Now that you've refactored the code, I think it's no longer needed, the code speaks for itself.

mapazarr replied:
Done.

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -319,3 +339,569 @@ def select_observations(self, selection=None):
+
+        return table
+
+    def table_to_axes(self, table):
```
Make this a `staticmethod`, to make it clear that `self` isn't used / modified.

```python
@staticmethod
def table_to_axes(table):
```
(and then change the call sites to call `ObservationGroups.table_to_axes`)

mapazarr replied:
Done.

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -319,3 +339,569 @@ def select_observations(self, selection=None):
+                    max_value = recover_units(self.obs_groups_table[name + '_MAX'][i_row],
+                                              self.obs_groups_table[name + '_MAX'])
+                elif format == 'bin_values':
+                   min_value = recover_units(self.obs_groups_table[name][i_row],
```
Use 4-space indentation.
(there's a few other PEP8 violations in your code ... it would be nice if you could run `pep8` and / or `pylint` and clean those up yourself)
(the 80 character line length limit doesn't have to be followed strictly ... if the code is more readable and has 85 lines, keep as is)

mapazarr replied:
Done.

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -319,3 +339,569 @@ def select_observations(self, selection=None):
+    @property
+    def info(self):
+        """Info string (str)"""
+        return ("{0} {1} {2}".format(self.name, self.format, self.bins))
```
Remove redundant parentheses.
Such things are pointed out by my editor, but you can also get a list by running pylint like this:
```bash
pylint gammapy/obs/observation.py
```

mapazarr replied:
Done.

mapazarr commented:
@cdeil OK, I addressed almost all comments. Please don't add any new ones, since I still need to finish a couple of thing (and sleep).

mapazarr commented:
@deil Ok, I finished with the comments. Can you please press the green button? There is surely room for improvement, but I can't stay awake longer.

cdeil commented:
@mapazarr – Merging this now. Thank you!

This is the biggest PR we've ever had in Gammapy (+3,682 −882). It contains a lot of great new functionality (observation handling, grouping, cube background model production).

One issue is that this goes in without substantial code review of methods, API, implementation, tests, docs because GSoC is over and I wasn't available much.
So we'll have to see how we handle this for the coming weeks / months. One possibility is that I take some time (a few days needed?) review the diff of this PR and the task list in the follow-up issue #333 to this PR.
Or we'll just do a big systematic review of all of Gammapy and try to improve it in the months before the 1.0 release ... probably we should do this as most code in Gammapy hasn't been thoroughly reviewed and many things changed (like became available in Astropy or we decided to use Sherpa).
Or a little bit of both.

cdeil commented:
The Sphinx build on RTD succeeded:
https://readthedocs.org/projects/gammapy/builds/3264248/
(and RTD since yesterday? shows a much better build log, with the commands they are running and the corresponding output)

Here's some of the new high-level docs pages from this PR ... they all look OK to me (formatting-wise ... I have to take a day next week to go through them):
* http://gammapy.readthedocs.org/en/latest/obs/observation_grouping.html
* http://gammapy.readthedocs.org/en/latest/background/index.html
* http://gammapy.readthedocs.org/en/latest/background/models.html
* http://gammapy.readthedocs.org/en/latest/background/make_models.html
* http://gammapy.readthedocs.org/en/latest/datasets/make_datasets.html

mapazarr commented:
![snapshot42](https://cloud.githubusercontent.com/assets/11326731/9414921/c60b7526-47f0-11e5-8f01-d32ba0b5fb5a.png)
Release Gammapy 0.3.1 ? :-D

mapazarr commented:
The work should continue in #333, or other subsequent PRs.

cdeil commented:
> Release Gammapy 0.3.1?

No, we'll just release 0.4 in ~ 1 month.
Better to get some usage / testing for the stuff from this PR. And I also want to have a first version of morphology and spectrum fitting in the next release (which is what most users want).
